keyboards/miiiw/blackio83/blackio83.c

```diff
@@ -0,0 +1,172 @@
+/* Copyright 2023 ArthurCyy <https://github.com/ArthurCyy>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "blackio83.h"
+#include "usb_main.h"
+#include "usb_util.h"
+
+#define LOOP_10HZ_PERIOD    100
+deferred_token loop10hz_token  = INVALID_DEFERRED_TOKEN;
+uint32_t loop_10Hz(uint32_t trigger_time, void *cb_arg);
+
+static const SerialConfig mwproto_uart_config = {
+    .speed = MWPROTO_BITRATE,
+};
+
+static void POWER_EnterSleep(void) {
+    /* Clear Wake-up flag */
+    PWR->CR |= PWR_CR_CWUF | PWR_CR_CSBF;
+    /* Select Sleep mode */
+    /* PWR->CR |= PWR_CR_PDDS | PWR_CR_LPDS; */
+    /* Set SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+    /* Request Wait For Interrupt */
+    __WFI();
+    /* Reset SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+    NVIC_SystemReset();
+}
+
+extern void ws2812_poweron(void);
+extern void ws2812_poweroff(void);
+
+void keyboard_pre_init_kb() {
+	keyboard_pre_init_user();
+
+    setPinInputLow(MWPROTO_STATUS_PIN);
+    setPinOutput(MWPROTO_WAKEUP_PIN);
+    writePinLow(MWPROTO_WAKEUP_PIN);
+    wait_ms(2);
+    writePinHigh(MWPROTO_WAKEUP_PIN);
+
+	palSetLineMode(MWPROTO_TX_PIN, PAL_MODE_ALTERNATE(MWPROTO_TX_PAL_MODE) | PAL_OUTPUT_TYPE_OPENDRAIN);
+	sdStart(&MWPROTO_DRIVER, &mwproto_uart_config);
+}
+
+void keyboard_post_init_kb(void) {
+    keyboard_post_init_user();
+
+    print(/* clang-format off */
+        "\n<--QMK Keyboard-->\n"
+        "Vendor: " STR(MANUFACTURER) "(" STR(VENDOR_ID) ")\n"
+        "Product: " STR(PRODUCT) " (" STR(PRODUCT_ID) ")\n"
+        "Version: " STR(DEVICE_VER) "\n"
+        "BUILD: " __DATE__ "\n"
+    ); /* clang-format on */
+
+    writePinLow(MWPROTO_WAKEUP_PIN);
+    wait_ms(50);
+    sdPutI(&MWPROTO_DRIVER, 0xA5);
+    sdPutI(&MWPROTO_DRIVER, 0x12);
+    sdPutI(&MWPROTO_DRIVER, 0x01);
+    sdPutI(&MWPROTO_DRIVER, 0x02);
+    sdPutI(&MWPROTO_DRIVER, 0xB4);
+    writePinHigh(MWPROTO_WAKEUP_PIN);
+
+    ws2812_poweron();
+    loop10hz_token = defer_exec(LOOP_10HZ_PERIOD, loop_10Hz, NULL);
+}
+
+void shutdown_user(void) {
+#ifdef RGB_MATRIX_ENABLE
+    rgb_matrix_set_suspend_state(true);
```

Comment:
Clever.  Though, this would only work if suspend is enabled for rgb matrix, in general.  If it's not... then it won't do anything.